lib/firenest/replicated_state.ex

```diff
@@ -0,0 +1,624 @@
+  can only be attached to local processes.
+
+  The state is managed through callbacks that are invoked on the
+  node where the process lives and remotely on other nodes when local
+  changes are propagated.
+
+  The state is replicated incrementally through building local "deltas"
+  (changes to the state) and periodically replicating them remotely. Only
+  some amount of recent deltas are retained for "catching up" remote nodes.
+  If remote node is too far behind the current state and a delta-based
+  catch-up can't be performed, the server falls back to transferring the entire
+  local state. If keeping track of incremental changes is not convenient
+  for a particular state type, the value of a delta can be set to be equal
+  to the current state - this will always cause full state transfers.
+  """
+  alias Firenest.SyncedServer
+
+  @type server() :: atom()
+  @type key() :: term()
+
+  @type local_delta() :: term()
+  @type remote_delta() :: term()
+  @type state() :: term()
+  @type config() :: term()
+
+  @type delayed_update() :: {timeout :: pos_integer(), update :: term()}
+
+  @doc """
+  Called when a partition starts up.
+
+  It returns the initial delta value used for incremental state change
+  tracking and  an immutable `config` value that will be passed to all
+  callbacks.
+  """
+  @callback init(opts :: keyword()) :: {initial_delta :: local_delta(), config()}
+
+  @doc """
+  Called whenever the `put/4` function is called to create a new state.
+
+  The `arg` is received from the corresponding `put/4` call. For the
+  explanation of the `delayed_update` and `:delete` return values see
+  the `c:local_update/4` callback.
+
+  This is a good place for broadcasting local state changes.
+  """
+  @callback local_put(arg :: term(), config()) ::
+              {:ok, initial_state} | {:ok, initial_state, delayed_update() | :delete}
+            when initial_state: state
+
+  @doc """
+  Called whenever the `update/4` function is called to update a state.
+
+  This is a good place for broadcasting local state changes.
+
+  ## Delayed update and delete
+
+  If the function returns a third element in the tuple consisting of
+  `{timeout, update}`, a timer is started by the server and the
+  `c:local_update/4` callback will be called with the `update` value
+  after `timeout` milliseconds.
+
+  If the third element is `:delete`, the state will be immediately
+  deleted and the `c:local_delete/2` callback triggered.
+  """
+  @callback local_update(update :: term(), local_delta(), state(), config()) ::
+              {local_delta(), state()} | {local_delta(), state(), delayed_update() | :delete}
+
+  @doc """
+  Called whenever attached process dies or the `delete/3` or `delete/2`
+  function is called and state is about to be deleted.
+
+  The return value is ignored.
+
+  This is a good place for broadcasting local state changes.
+  """
+  @callback local_delete(state(), config()) :: term()
```

Comment:
@michalmuskala I'm wondering how do you plan to implement the removal of the part of state bound to terminating process from remote nodes (and local node as well: if I understand this correctly there might be many processes on a single note contributing to the state under the key) ? Unless I'm missing something, there will need to be a way to produce "undo" deltas in those situations, right?

Oh, and sorry if it's too soon to ask those questions, I'm simply curious 😄

Reply:
No, those questions are great 😃 It allows verifying the design.

`local_delete/2` is called only in response to the `delete` function in the local node. Remotely, if you want to observer removes, etc, you'd use the `observe_remote_changes` callback. There you might get:
```elixir
[
  {:foo, [{state_node1, [:delete]}, {state_node2, [{:delta, delta}]}]},
  {:bar, [{state_node1, [:delete]}]
]
```
Which means key `:foo` and `:bar` were removed on node 1 and key `:foo` was updated at node 2.

Reply:
Ah, I get it. So by default there won't be any wiring for tracking removed parts of state, and we would need to do that by implementing `observe_remote_changes`.